### PR TITLE
Improve question validation errors

### LINKED ISSUE #618: Improve question validation
- Add err messages for question inputs
- Blend questions block validation UI

### PR DESCRIPTION
**Pull request by borna223**
Resolves #618

### CODE REVIEW

File: client/utils/assessment.js [hunk 1]
```diff
@@ -209,3 +91,7 @@ export const defaults = {
     return element;
   }
 };
+
+export const getErrorMessages = (errors, pattern) => errors
```
Comment (underscope) [inline client/utils/assessment.js]: * pattern

File: client/components/content-elements/tce-multiple-choice/edit/index.vue [hunk 2]
```diff
@@ -25,18 +25,22 @@
...
+          {{ correctError }}
+        </div>
+      </transition>
       <v-btn
         v-if="isEditing"
         @click="addAnswer"
         :color="color"
-        text
-        class="px-2">
+        text class="ml-auto px-2">
```
Comment (underscope) [inline client/components/content-elements/tce-multiple-choice/edit/index.vue]: Move class to the next line

File: client/components/content-elements/tce-multiple-choice/edit/index.vue [hunk 3]
```diff
@@ -131,4 +128,9 @@ export default {
 .v-avatar {
   color: #fff;
 }
+
+.correct-error {
```
Comment (underscope) [inline client/components/content-elements/tce-multiple-choice/edit/index.vue]: Extrapolate in the question wrapper (reused...) or extract the whole component like mentioned before.

File: client/components/common/tce-core/QuestionContainer/Question.vue [hunk 4]
```diff
@@ -1,25 +1,30 @@
...
+    <div class="question">
       <draggable v-model="question" v-bind="dragOptions" class="row">
         <contained-content
           v-for="element in question" :key="element.id"
           @save="updateElement(element, $event)"
           @delete="deleteElement(element)"
           :element="element"
           :is-disabled="!isEditing"
-          dense />
+          dense class="mb-4" />
```
Comment (underscope) [inline client/components/common/tce-core/QuestionContainer/Question.vue]: Class attr should go to the next line

File: client/components/common/tce-core/QuestionContainer/Question.vue [hunk 5]
```diff
@@ -1,25 +1,30 @@
...
-      <add-element
-        @add="addElement"
-        :layout="false"
-        :disabled="!isEditing"
-        :include="['JODIT_HTML', 'IMAGE', 'EMBED', 'HTML']"
-        :class="['add-element', { invisible: !isEditing }]"
-        label="Add question block"
-        large />
     </div>
+    <transition name="message-transition">
```
Comment (underscope) [inline client/components/common/tce-core/QuestionContainer/Question.vue]: Since this one is being reused (similar needed elsewhere), consider wrapping it into tce-core common (question) component

File: client/components/content-elements/tce-single-choice/edit/index.vue [hunk 6]
```diff
@@ -1,15 +1,15 @@
 <template>
   <div>
     <div class="subtitle-2">{{ title }}</div>
-    <v-radio-group v-model="correct" :error="correctError" hide-details>
+    <v-radio-group v-model="correct" :errors="!!correctError" hide-details>
```
Comment (underscope) [inline client/components/content-elements/tce-single-choice/edit/index.vue]: error ? => https://vuetifyjs.com/en/api/v-radio-group/

File: client/components/content-elements/tce-single-choice/edit/index.vue [hunk 7]
```diff
@@ -18,19 +18,23 @@
...
+          {{ correctError }}
+        </div>
+      </transition>
       <v-btn
         v-if="isEditing"
         @click="addAnswer"
         :color="color"
-        text
-        class="px-2">
+        text class="ml-auto px-2">
```
Comment (underscope) [inline client/components/content-elements/tce-single-choice/edit/index.vue]: class should go to the next line

File: client/components/content-elements/tce-fill-blank/edit/index.vue [hunk 8]
```diff
@@ -75,12 +80,8 @@ import times from 'lodash/times';
 const TEXT_TYPES = ['JODIT_HTML', 'HTML'];
 const BLANK = /(@blank)/g;
 
-const OUT_OF_SYNC_ERR = {
-  type: 'error',
-  text: `
-    Question and blanks are out of sync!
-    Please delete unnecessary answers or add blanks in the question!`
-};
+const SYNC_ERROR = `Question and blanks are out of sync! Please delete
```
Comment (underscope) [inline client/components/content-elements/tce-fill-blank/edit/index.vue]: Start message on the new line
```
const SYNC_ERROR = `
  Question and.....`;
```

File: client/components/common/tce-core/InputError.vue [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+export default {
+  props: {
+    error: { type: String, default: '' }
+  }
+};
+</script>
+
+<style lang="scss" scoped>
+.input-error {
+  color: #ff5252;
```
Comment (underscope) [inline client/components/common/tce-core/InputError.vue]: This is red accent-2 so you can pull it up as `var(--v-red-accent2)`

File: client/components/content-elements/tce-single-choice/edit/index.vue [hunk 10]
```diff
@@ -18,19 +18,20 @@
...
         <template slot="append">
-          <v-btn v-if="isEditing" @click="removeAnswer(idx)" small icon>
+          <v-btn v-if="allowRemoval" @click="removeAnswer(idx)" small icon>
             <v-icon small>mdi-close</v-icon>
           </v-btn>
         </template>
       </v-text-field>
     </v-radio-group>
-    <div :class="['d-flex', 'justify-end', { 'pb-2': isEditing }]">
+    <div :class="['d-flex justify-space-between align-center', { 'pb-2': isEditing }]">
```
Comment (underscope) [inline client/components/content-elements/tce-single-choice/edit/index.vue]: Consider separating dynamic bindings, this is a bit lengthy.

File: client/components/content-elements/tce-multiple-choice/edit/index.vue [hunk 11]
```diff
@@ -25,18 +25,19 @@
...
       </template>
       <template slot="append">
-        <v-btn v-if="isEditing" @click="removeAnswer(idx)" small icon>
+        <v-btn v-if="allowRemoval" @click="removeAnswer(idx)" small icon>
           <v-icon small>mdi-close</v-icon>
         </v-btn>
       </template>
     </v-text-field>
-    <div :class="['d-flex', 'justify-end', { 'pb-2': isEditing }]">
+    <div :class="['d-flex justify-space-between align-center', { 'pb-2': isEditing }]">
```
Comment (underscope) [inline client/components/content-elements/tce-multiple-choice/edit/index.vue]: * See single-choice comment

File: client/components/content-elements/tce-fill-blank/edit/index.vue [hunk 12]
```diff
@@ -106,11 +106,11 @@ export default {
     },
     question: vm => vm.assessment.question,
     disabled: vm => !vm.isEditing,
-    hasAnswers: vm => !!size(vm.correct),
     blankCount: vm => getBlankCount(vm.question),
     blankCountInfo: vm => getCountInfo(vm.blankCount),
     isSynced: vm => vm.blankCount === size(vm.correct),
-    color: vm => vm.disabled ? 'grey' : 'grey darken-3'
+    color: vm => vm.disabled ? 'grey' : 'grey darken-3',
+    syncError: vm => (vm.isGraded && !vm.isSynced && SYNC_ERROR) || ''
```
Comment (underscope) [inline client/components/content-elements/tce-fill-blank/edit/index.vue]: Ternary is more appropriate/readable in this case 😉

File: client/components/common/tce-core/QuestionContainer/index.vue [hunk 13]
```diff
@@ -92,24 +89,23 @@ export default {
...
-        ...this.isGraded ? elementSchema : omit(elementSchema, ['correct'])
+        ...this.isGraded ? schema : omit(schema, ['correct'])
       });
     },
-    hasErrors: vm => !!vm.errors.length,
-    answerType: vm => vm.element.data.type,
-    isGraded: vm => vm.element.type === 'ASSESSMENT',
-    showFeedback: vm => WITH_FEEDBACK.includes(vm.answerType),
-    hintError: vm => vm.errors.includes('hint')
+    hintErrors() {
```
Comment (underscope) [inline client/components/common/tce-core/QuestionContainer/index.vue]: Can be inlined like before 💅

File: client/components/common/tce-core/QuestionContainer/Question.vue [hunk 14]
```diff
@@ -1,34 +1,54 @@
...
+      :include="['JODIT_HTML', 'IMAGE', 'EMBED', 'HTML']"
+      :class="{ invisible: !isEditing }">
+      <div class="d-flex justify-space-between mt-2 pl-3">
+        <input-error :error="questionError" />
+        <v-btn
+          @click="addElement"
+          text
+          class="ml-auto px-2 mt-2">
+          <v-icon dense class="mr-1">mdi-plus</v-icon>
+          Add question
```
Comment (underscope) [inline client/components/common/tce-core/QuestionContainer/Question.vue]: Add question element

File: client/components/common/tce-core/AddElement/index.vue [hunk 15]
```diff
@@ -1,21 +1,23 @@
...
+        @click.stop="showElementPicker"
+        color="blue-grey darken-3"
+        text
+        class="mt-3 mb-4">
+        <v-icon class="pr-2">{{ icon }}</v-icon>{{ label }}
+      </v-btn>
+      <v-btn
+        v-else
+        @click.stop="showElementPicker"
+        icon
```
Comment (underscope) [inline client/components/common/tce-core/AddElement/index.vue]: Inline booleans

File: client/components/common/tce-core/AddElement/index.vue [hunk 16]
```diff
@@ -1,21 +1,23 @@
...
+        text
+        color="blue-grey darken-3">
```
Comment (underscope) [inline client/components/common/tce-core/AddElement/index.vue]: Order

File: client/components/common/tce-core/AddElement/index.vue [hunk 17]
```diff
@@ -1,21 +1,23 @@
...
-    </v-btn>
-    <v-btn
-      v-else
-      @click.stop="isVisible = true"
-      icon
-      text
-      color="blue-grey darken-3">
-      <v-icon>{{ icon }}</v-icon>
-    </v-btn>
+    <slot :addElement="showElementPicker">
```
Comment (underscope) [inline client/components/common/tce-core/AddElement/index.vue:3]: Why did you need to add a slot? / what did you gained with method extraction?

File: client/components/common/tce-core/QuestionContainer/Question.vue [hunk 18]
```diff
@@ -78,39 +100,82 @@ export default {
       this.$emit('update', { question });
     }
   },
+  created() {
+    this.editorChannel.on('element:focus', (element = {}) => {
+      this.isFocused = !!find(this.question, q => q.id === element.id);
```
Comment (underscope) [inline client/components/common/tce-core/QuestionContainer/Question.vue]: Use find shorthand.

File: client/components/common/tce-core/QuestionContainer/Question.vue [hunk 19]
```diff
@@ -78,39 +100,82 @@ export default {
...
+      to right,
+      $disabled-color 0,
+      $disabled-color 0.125rem,
+      transparent 0.125rem,
+      transparent 0.25rem
+    ) 1 repeat;
+}
+
+.incorrect::before, .incorrect::after {
+  border-color: #ff5252;
```
Comment (underscope) [inline client/components/common/tce-core/QuestionContainer/Question.vue]: See prev color-related comment

File: client/components/common/tce-core/QuestionContainer/Question.vue [hunk 20]
```diff
@@ -78,39 +100,82 @@ export default {
...
+$disabled-color: rgba(0, 0, 0, 0.38);
+$swing: cubic-bezier(0.25, 0.8, 0.5, 1);
+
 .question-container {
-  clear: both;
-  margin: 0 1.5rem;
+  position: relative;
+  min-height: 8.75rem;
+  padding: 1rem 1.75rem 0;
+  border-radius: 0.125rem;
```
Comment (underscope) [inline client/components/common/tce-core/QuestionContainer/Question.vue:127]: Move after background

File: client/components/common/tce-core/AddElement/index.vue (hunk 17 above)
Comment (borna223) [reply client/components/common/tce-core/AddElement/index.vue:3]: The goal was to make the `Add question element` button stylistically identical to the `Add answer` button (SC, MC, FB, NR...) and position it correctly where needed. 

`tce-core/add-element` is most likely reused within some custom containers and/or elements so I didn't want to change it. Adding a slot with fallback content seemed better choice than styling it with v-deep from `tce-core/question-container/question`.

File: client/components/common/tce-core/InputError.vue [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+export default {
+  props: {
+    error: { type: String, default: '' }
+  }
+};
+</script>
+
+<style lang="scss" scoped>
+.input-error {
+  color: var(--v-error-base);
```
Comment (underscope) [inline client/components/common/tce-core/InputError.vue:19]: 🏆

File: client/components/common/tce-core/QuestionContainer/Question.vue [hunk 22]
```diff
@@ -54,11 +76,11 @@ export default {
         this.$emit('update', { question });
       }
     },
-    questionError: vm => vm.errors.includes('question'),
+    questionError: vm => head(getErrorMessages(vm.errors, 'question')),
     dragOptions: () => DRAG_OPTIONS
   },
   methods: {
-    addElement(element) {
+    addQuestion(element) {
```
Comment (underscope) [inline client/components/common/tce-core/QuestionContainer/Question.vue]: addQuestionElement (Same logic as for label ;) )

**Comment by underscope** [comment]
@kjuej Merging this, please revisit on develop 😉
